src/Sentry/Internal/MemoryMonitor.cs

```diff
@@ -0,0 +1,164 @@
+{
+    private readonly long _totalMemory;
+
+    private readonly SentryOptions _options;
+    private readonly HeapDumpTrigger _dumpTrigger;
+
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+    private readonly Action _onCaptureDump; // Just for testing purposes
+    private readonly Action<string> _onDumpCollected;
+
+    public MemoryMonitor(SentryOptions options, Action<string> onDumpCollected, Action? onCaptureDump = null)
+    {
+        _options = options;
+        _dumpTrigger = options.HeapDumpTrigger
+                       ?? throw new ArgumentException("No heap dump trigger configured on the options", nameof(options));
+        _onDumpCollected = onDumpCollected;
+        _onCaptureDump = onCaptureDump ?? CaptureMemoryDump;
+
+        _totalMemory = GC.GetGCMemoryInfo().TotalAvailableMemoryBytes;
+
+        GarbageCollectionMonitor.Start(CheckMemoryUsage, _cancellationTokenSource.Token);
+    }
+
+    internal void CheckMemoryUsage()
+    {
+        var eventTime = DateTimeOffset.UtcNow;
+        if (!_options.HeapDumpDebouncer.CanProcess(eventTime))
+        {
+            return;
+        }
+
+        var usedMemory = Environment.WorkingSet;
+        if (!_dumpTrigger(usedMemory, _totalMemory))
+        {
+            return;
+        }
+
+        _options.HeapDumpDebouncer.RecordOccurence(eventTime);
+
+        var usedMemoryPercentage = ((double)usedMemory / _totalMemory) * 100;
+        _options.LogDebug("Total Memory: {0:N0} bytes", _totalMemory);
+        _options.LogDebug("Memory used: {0:N0} bytes ({1:N2}%)", usedMemory, usedMemoryPercentage);
+        _options.LogDebug("Automatic heap dump triggered");
+        _onCaptureDump();
+    }
+
+    internal void CaptureMemoryDump()
+    {
+        if (_options.DisableFileWrite)
+        {
+            _options.LogDebug("File write has been disabled via the options. Unable to create memory dump.");
+            return;
+        }
+
+        var dumpFile = TryGetDumpLocation();
+        if (dumpFile is null)
+        {
+            return;
+        }
+
+        var processId = Environment.ProcessId;
+        _options.LogInfo("Creating a memory dump for Process ID: {0}", processId);
+
+        // Check which patth to use for dotnet-gcdump. If it's been bundled with the application, it will be available
+        // in the `dotnet-gcdump` folder of the application directory. Otherwise we assume it has been installed globally.
+        var bundledToolPath = Path.Combine(AppContext.BaseDirectory, "dotnet-gcdump", "dotnet-gcdump.dll");
+        if (File.Exists(bundledToolPath))
+        {
+            _options.LogDebug($"Using bundled version of dotnet-gcdump from: {bundledToolPath}");
+        }
+        else
+        {
+            _options.LogDebug("Using global version of dotnet-gcdump");
+        }
+
+        var arguments = $"collect -p {processId} -o '{dumpFile}'";
+        var command = File.Exists(bundledToolPath)
```

Comment:
Spare 1 file I/O by assigning the return of the first check 

And, lets always use `_options.FileSystem.FileExists`. 
Ideally we'd have an analyzer because I only realized the `File.Exists` usage because I saw the second pattern further down.

* https://github.com/getsentry/sentry-dotnet/issues/3643

```suggestion
        var bundledFileExists = _options.FileSystem.FileExists(bundledToolPath);
        if (bundledFileExists)
        {
            _options.LogDebug($"Using bundled version of dotnet-gcdump from: {bundledToolPath}");
        }
        else
        {
            _options.LogDebug("Using global version of dotnet-gcdump");
        }

        var arguments = $"collect -p {processId} -o '{dumpFile}'";
        var command = bundledFileExists
            ? $"dotnet {bundledToolPath} {arguments}"
            : $"dotnet-gcdump {arguments}";
```